### PR TITLE
Add support for overloaded signatures and add overloaded ActiveRecordRelations sigs

### LINKED ISSUE #1789: Is there a way to write a compiler that defines multiple `sig` for a single method?
For example for `T::Array` there are multiple overloads/sigs for `first` and `first(limit)`. I'd like to be able to do the same with a compiler.

There's `create_method`. I tried calling it twice with the same method name but that ends up creating two methods with two sigs, and Sorbet takes the last definition. I couldn't find anything in the API to do this... but maybe it's possible?

### PR DESCRIPTION
**Pull request by bdewater**
### Motivation
Close #1789

### Implementation
Based on the patch in aforementioned issue.

### Tests
Updated tests.

### CODE REVIEW

**Comment by KaanOzkan** [comment]
@bdewater it's resolved on main, you can rebase.

File: lib/tapioca/rbi_ext/model.rb [hunk 1]
```diff
@@ -87,24 +87,59 @@ def create_type_variable(name, type:, variance: :invariant, fixed: nil, upper: n
...
+        return_type: String,
+      ).returns(RBI::Sig)
+    end
+    def create_sig(parameters: [], return_type: "T.untyped")
+      sig = RBI::Sig.new(return_type: return_type)
+      parameters.each do |param|
+        sig << RBI::SigParam.new(param.param.name, param.type)
+      end
+      sig
+    end
```
Comment (paracycle) [inline lib/tapioca/rbi_ext/model.rb:141]:
```suggestion
    def create_sig(parameters: [], return_type: "T.untyped")
      params = parameters.map do |param|
        RBI::SigParam.new(param.param.name, param.type)
      end
      
      RBI::Sig.new(params: params, return_type: return_type)
    end
```

**Comment by bdewater** [comment]
Thanks for the feedback @paracycle! Rebased and incorporated your suggestions.

**Review by paracycle — APPROVED** [review]
Thank you!

**Comment by bdewater** [comment]
I fixed the last commit ([here](https://github.com/bdewater/tapioca/commit/4457b30d82b5f3b51d5dbc8c0467202cb2c00811)) but it doesn't show up yet because of https://www.githubstatus.com/incidents/wcl1sw4mzg60

File: lib/tapioca/dsl/compilers/active_record_relations.rb [hunk 2]
```diff
@@ -560,12 +550,25 @@ def create_common_methods
...
+              end
+              sigs = [
+                common_relation_methods_module.create_sig(
+                  parameters: [create_param("args", type: args_type)],
+                  return_type: "T::Enumerable[#{constant_name}]",
+                ),
+                common_relation_methods_module.create_sig(
+                  parameters: [create_param("args", type: "T.untyped")],
+                  return_type: constant_name,
+                ),
```
Comment (KaanOzkan) [inline lib/tapioca/dsl/compilers/active_record_relations.rb:566]: This pattern doesn't seem to work
```ruby 
# test.rbi
class Foo
  sig { returns(T.untyped) }
  def foo; end

  sig { returns(T::Array[T.untyped]) }
  def bar; end
end

sig { params(args: T::Array[T.untyped]).returns(T::Enumerable[Foo]) }
sig { params(args: T.untyped).returns(Foo) }
def find(args); end
```

```ruby
# test.rb
T.reveal_type(find(Foo.new.foo)) # T::Enumerable[Foo]
T.reveal_type(find(Foo.new.bar)) # T::Enumerable[Foo]
```

First sig always takes precedence. I'm not sure why, maybe it's a sorbet bug. We could open a ticket and revert this change until its fixed.

Comment (bdewater) [reply lib/tapioca/dsl/compilers/active_record_relations.rb:566]: T.untyped argument would be matched against the first signature since it's a subtype of everything, as Jez pointed out in the linked Sorbet issue. 

Also noticed this particular issue was documented on https://sorbet.org/docs/overloads 😅 
> In the presence of untyped arguments, chances are high that the first overload is selected, which might not be desired.

Comment (bdewater) [reply lib/tapioca/dsl/compilers/active_record_relations.rb]: Some experimenting with the Rails console - `find(nil)` errors [here](https://github.com/rails/rails/blob/v7.1.3.2/activerecord/lib/active_record/relation/finder_methods.rb#L496-L497), `find({})` errors [here](https://github.com/rails/rails/blob/v7.1.3.2/activerecord/lib/active_record/connection_adapters/abstract/quoting.rb#L14-L28) which gives a pretty good list of types that can be passed.

File: lib/tapioca/dsl/compilers/active_record_relations.rb [hunk 3]
```diff
@@ -560,12 +550,29 @@ def create_common_methods
...
+              sigs = [
+                common_relation_methods_module.create_sig(
+                  parameters: [create_param("args", type: id_types)],
+                  return_type: constant_name,
+                ),
+                common_relation_methods_module.create_sig(
+                  parameters: [create_param("args", type: array_type)],
+                  return_type: "T::Enumerable[#{constant_name}]",
+                ),
+              ]
```
Comment (amomchilov) [inline lib/tapioca/dsl/compilers/active_record_relations.rb:571]: This is missing one overload:

`find` can be passed a block, in which case it delegates to `Enumerable#find`, which returns a nilable result.

**Comment by alexevanczuk** [comment]
@KaanOzkan Would you be able to cut a new release that includes this? Would love to add this to add overloads for some RBI generation we have today.

**Comment by KaanOzkan** [comment]
@alexevanczuk I'm not sure what you mean, this has been released since [v0.13.0](https://github.com/Shopify/tapioca/releases/tag/v0.13.0). You should be seeing  overloaded sigs under `CommonRelationMethods` in a model RBI, for example the `find` method.
